network/metadata_cache.go

```diff
@@ -0,0 +1,59 @@
+package network
+
+import (
+	"github.com/grafana/walqueue/types"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+type cachedMetadata struct {
+	SendAttempted bool
+	Help          string
+	Type          prompb.MetricMetadata_MetricType
+	Unit          string
+}
+
+type metadataCache struct {
+	items *lru.Cache[string, cachedMetadata]
+}
+
+func NewMetadataCache(size int) (*metadataCache, error) {
+	cache, err := lru.New[string, cachedMetadata](size)
+	if err != nil {
+		return nil, err
+	}
+	return &metadataCache{
+		items: cache,
+	}, nil
+}
+
+func (c *metadataCache) GetIfNotSent(key string) (cachedMetadata, bool) {
```

Comment:
Do we need to decouple the Get / Sent logic here? The assumption is if we get it from the cache we have sent it but it's possible we get it and the send fails right?

Reply:
IMO this can be a followup investigation.